### ISSUE TITLE
[Bug Report]: Crashing on Atmosphere 1.4.0 Hekate 5.9.0 HOS 15.0

### ISSUE THREAD

**Issue by CantWeAllDisagree** [ndeadly/MissionControl#519]
### Switch Firmware Version

Other (please specify in issue description)

### Atmosphère Version

Other (please specify in issue description)

### Mission Control Version

0.7.1 (Latest)

### Boot Method

Fusée

### Issue Description

Newest Firmware Update causes atmosphere to crash when launching mission control. Using new Hekate bootloader 5.9.0 and atmosphere 1.4.0. Needs update for HOS 15.0

**Comment by Masamune3210** [comment]
Bluetooth module got a update, gonna be a bit prob before MissionControl gets support depending on how much was changed

**Comment by ndeadly** [comment]
As far as I can tell, nothing has broken with the updates to bluetooth. Mostly waiting for Lockpick_RCM to be updated so I can obtain keys to decrypt the new firmware and find offsets for generating bluetooth patches.

**Comment by ndeadly** [comment]
Offsets have been located and tentative code for the patches is now written. The problem? Now we don't have enough unused bytes in the .text padding to patch them in 😞 Will have to get a bit creative with this one.

**Comment by gucciboy512** [comment]
this is also broken on 15.0.0 and latest atmosphere

**Comment by ndeadly** [comment]
> this is also broken on 15.0.0 and latest atmosphere

What do you think this thread is for? 🤨

**Comment by ndeadly** [comment]
I made a pre-release to support 15.0.0 while I work on the patches.

https://github.com/ndeadly/MissionControl/releases/tag/v0.8.0

**Comment by ndeadly** [comment]
Here are some preliminary patches for 15.0.0. Could I get some feedback on whether they're working for Wii/WiiU and Xbox Elite 2 controllers? In particular the XBE2, as I don't own one. Seems to be fine with my Wii controllers.

[mc_1500_exefs_patches.zip](https://github.com/ndeadly/MissionControl/files/9892518/mc_1500_exefs_patches.zip)

**Comment by alint77** [comment]
it does not work with XBE2

**Comment by Tatsuhik-o** [comment]
> Here are some preliminary patches for 15.0.0. Could I get some feedback on whether they're working for Wii/WiiU and Xbox Elite 2 controllers? In particular the XBE2, as I don't own one. Seems to be fine with my Wii controllers.
> 
> [~mc_1500_exefs_patches.zip~](https://github.com/ndeadly/MissionControl/files/9892518/mc_1500_exefs_patches.zip) [mc_1500_exefs_patches_2.zip](https://github.com/ndeadly/MissionControl/files/9895787/mc_1500_exefs_patches_2.zip)

I have tried with the pre-release and patch mentioned , sadly neither the DS4 nor the elite series core worked. the light on both of them sit still as if it were connected then the controller shut down.

**Comment by ndeadly** [comment]
@Tatsuhik-o the Elite Series Core is the new one right? It's very likely a Bluetooth LE controller like the Series X|S ones and can't be downgraded to something that works. It might even have a new vid/pid that would need to be whitelisted too. Can you pair it under Windows and check the device properties?

As for the DS4 not working, that will be another matter entirely. You can search through the issue tracker for recent threads about PS4 controllers. There are several suggestions you can try

**Comment by Tatsuhik-o** [comment]
> 

I have checked in microsoft website about the elite series core and it does only support Bluetooth LE, so sorry for not being able to test this one.

About DS4, I tried both the motion calibration build as well as emptying the controller folder in config

**Comment by ndeadly** [comment]
Yeah, I think the patches still aren't right and are compromising other controllers. I was testing Wii controller patches at the same time and since they were working assumed all was good, forgetting they use a different pairing method. Try removing the patch file (and rebooting) and confirm the DS4 works.

**Comment by Tatsuhik-o** [comment]
> compromising

Still same issue with the DS4, I am using the no motion calibration build though

**Comment by ndeadly** [comment]
Oh yeah, duh, you still need device class patches on the new firmwares. Try this one

[mc_1500_exefs_patches_no_xbe2.zip](https://github.com/ndeadly/MissionControl/files/9896878/mc_1500_exefs_patches_no_xbe2.zip)

**Comment by Tatsuhik-o** [comment]
> Oh yeah, duh, you still need device class patches on the new firmwares. Try this one
> 
> [mc_1500_exefs_patches_no_xbe2.zip](https://github.com/ndeadly/MissionControl/files/9896878/mc_1500_exefs_patches_no_xbe2.zip)

I am still facing the same issue, tried to remove all patches together and try but no luck. I even tried syscon to see if my controller will connect via cable and it seems to be working fine.

**Comment by ndeadly** [comment]
My DS4 works again with the patches posted above, so I think this is a problem on your end. Try resetting the controller or deleting the `/config/MissionControl/controllers` directory.

**Comment by Tatsuhik-o** [comment]
> My DS4 works again with the patches posted above, so I think this is a problem on your end. Try resetting the controller or deleting the `/config/MissionControl/controllers` directory.

I still have no luck even after deleting the controller folder, I think i should revert back to the normal pre-release and try again

**Comment by JacobWiebe** [comment]
Confirming XBE2 works on 15.0.1, latest Atmosphere w/ your prerelease :^)

Cheers to you

**Comment by ndeadly** [comment]
@JacobWiebe that's what I wanted to hear :)

Just to be clear, you did try to pair the controller at the change grip/order screen rather than just reconnecting one you had already paired previously? 

And, since there are multiple different revisions of the patches floating around in the GitHub issues, could you please clarify which one you installed that worked for you?

**Comment by corybsa** [comment]
@ndeadly I also got my XBE2 working. Controller's firmware is 4.8.1908.0 (which shipped with the controller). I remember previous versions of Mission Control was not compatible with the latest controller firmware, does this patch allow me to update the controller firmware and still connect?

I did pair through the "Change grip/order" screen
HOS - 15.0.0
Atmosphere - 1.4.0
Hekate - 5.9.0
MissonControl - 0.8.0
Patch used - [mc_1500_exefs_patches_2.zip](https://github.com/ndeadly/MissionControl/files/9895787/mc_1500_exefs_patches_2.zip)

**Comment by ndeadly** [comment]
@corybsa awesome! Assuming you actually put the controller into sync mode rather than just reconnecting, it sounds like it's working. 

> does this patch allow me to update the controller firmware and still connect?

No, Bluetooth LE is still unsupported. These patches are required for regular bluetooth pairing on the older firmwares. The XBE2 sends an SDP record containing a _huge_ hid report descriptor that is larger than the buffer allocated in the bluetooth stack for storing such records. This causes the stack to abort authentication with the controller. Since the switch doesn't even use the descriptor in the end anyway, I just patch in a dummy one of minimal size in its place.

**Comment by corybsa** [comment]
@ndeadly Just to be thorough I wiped my bluetooth pairing database just now and tried to reconnect. Looks like it is not pairing now. Are there other patches I can try? Or is there anything I can do to help troubleshoot this?

**Comment by ndeadly** [comment]
@corybsa I posted another set here https://github.com/ndeadly/MissionControl/issues/3#issuecomment-1296378017

They're not actually any different though, I just swapped the ordering of the Wii and XBE2 patches in case there were any bugs in my script that lays the code out in memory. I've looked over the patch placement in memory and everything looks good, so I suspect the result will be the same, but it can't hurt to try. Don't forget to reboot after installing the new patch file.
